File: samples/cpp/page_cache.cpp
/*
 *  page_cache.cpp — companion sample for page-cache-coherency.md.
 *
 *  Runs the same hot-page ping-pong twice and lets MON$IO_STATS tell the
 *  difference between the two cache topologies:
 *
 *    phase 1 — two client processes -> ONE SuperServer shared cache
 *              (coherency by shared memory; almost no physical I/O)
 *    phase 2 — two EMBEDDED engine processes with PRIVATE caches over one
 *              file (ServerMode=SuperClassic sandbox; coherency by LCK_bdb
 *              page locks + blocking ASTs — data travels through the disk,
 *              so the same workload turns into hundreds of reads AND writes)
 *
 *  The parent only forks/execs; every attachment lives in a child process
 *  (each embedded child IS a full engine).  Rows 1 and 2 share a data page,
 *  so the two writers fight over one page without ever touching one row.
 */

#include "fb_sample.h"
#include <sys/stat.h>
#include <sys/wait.h>
#include <unistd.h>

using namespace fbsample;

static const char* SRV_DB = "inet://localhost//tmp/fbhandson/page_cache_srv.fdb";
static const char* EMB_DB = "/tmp/fbhandson/page_cache_emb.fdb";
static const char* SANDBOX = "/tmp/fbhandson/fbemb";
static const int ROUNDS = 300;

static int initDb(const char* conn)          // --init: fresh table, two rows
{
	Db db;
	db.attachOrCreate(conn);
	ITransaction* t = db.start();
	try { db.exec(t, "drop table t"); } catch (const FbException&) {}
	t->commit(&db.status);
	t = db.start();
	db.exec(t, "create table t (id int primary key, v int)");
	t->commit(&db.status);
	t = db.start();
	db.exec(t, "insert into t values (1, 0)");
	db.exec(t, "insert into t values (2, 0)");
	t->commit(&db.status);
	return 0;
}

static int worker(const char* conn, const char* rowId)   // --worker
{
	Db db;
	db.attach(conn);
	for (int i = 0; i < ROUNDS; ++i)
	{
		ITransaction* t = db.start();
		db.exec(t, std::string("update t set v = v + 1 where id = ") + rowId);
		t->commit(&db.status);
	}
	ITransaction* t = db.start();
	Db::Table io = db.query(t,
		"select MON$PAGE_FETCHES, MON$PAGE_READS, MON$PAGE_WRITES "
		"from MON$IO_STATS join MON$ATTACHMENTS using (MON$STAT_ID) "
		"where MON$ATTACHMENT_ID = CURRENT_CONNECTION");
	printf("  worker pid %-6d row %s: %d commits | page fetches=%-6s "
		"reads=%-4s writes=%s\n", getpid(), rowId, ROUNDS,
		io.rows[0][0].c_str(), io.rows[0][1].c_str(), io.rows[0][2].c_str());
	t->commit(&db.status);
	return 0;
}

static int check(const char* conn)           // --check: any lost updates?
{
	Db db;
	db.attach(conn);
	ITransaction* t = db.start();
	Db::Table r = db.query(t, "select id, v from t order by id");
	for (auto& row : r.rows)
		printf("  final: id=%s v=%s (expected %d)\n",
			row[0].c_str(), row[1].c_str(), ROUNDS);
	t->commit(&db.status);
	return 0;
}

static void run(const char* self, std::initializer_list<const char*> args)
{
	if (fork() == 0)
	{
		std::vector<const char*> v{self};
		v.insert(v.end(), args.begin(), args.end());
		v.push_back(nullptr);
		execv(self, const_cast<char**>(v.data()));
		_exit(127);
	}
}

static void await() { while (wait(nullptr) > 0) {} }

int main(int argc, char** argv)
{
	try
	{
		if (argc > 2 && !strcmp(argv[1], "--init"))   return initDb(argv[2]);
		if (argc > 3 && !strcmp(argv[1], "--worker")) return worker(argv[2], argv[3]);
		if (argc > 2 && !strcmp(argv[1], "--check"))  return check(argv[2]);

		printf("phase 1: two client processes, ONE SuperServer shared cache\n");
		fflush(stdout);
		run(argv[0], {"--init", SRV_DB}); await();
		run(argv[0], {"--worker", SRV_DB, "1"});
		run(argv[0], {"--worker", SRV_DB, "2"});
		await();
		run(argv[0], {"--check", SRV_DB}); await();

		// Build the embedded sandbox: a FIREBIRD root whose firebird.conf
		// says SuperClassic, so each process locks the file SHARED and runs
		// its own page cache (see 'Three layers of arbitration').
		mkdir(SANDBOX, 0777);
		for (const char* f : {"plugins", "intl", "tzdata", "firebird.msg", "security6.fdb"})
			symlink((std::string("/opt/firebird/") + f).c_str(),
				(std::string(SANDBOX) + "/" + f).c_str());
		FILE* conf = fopen((std::string(SANDBOX) + "/firebird.conf").c_str(), "w");
		fputs("ServerMode = SuperClassic\n", conf);
		fclose(conf);
		setenv("FIREBIRD", SANDBOX, 1);
		unlink(EMB_DB);

		printf("phase 2: two EMBEDDED engine processes, PRIVATE page caches\n");
		fflush(stdout);
		run(argv[0], {"--init", EMB_DB}); await();
		run(argv[0], {"--worker", EMB_DB, "1"});
		run(argv[0], {"--worker", EMB_DB, "2"});
		await();
		run(argv[0], {"--check", EMB_DB}); await();
		printf("same workload — the private caches paid for coherency in disk I/O.\n");
		return 0;
	}
	catch (const FbException& e) { return report(e); }
}

File: samples/rust/src/bin/page_cache.rs
//! page_cache.rs — one shared cache vs. private caches over one file.
//!
//! The rsfbclient twin of ../cpp/page_cache.cpp (see
//! ../../page-cache-coherency.md).  The same hot-page ping-pong runs twice
//! and MON$IO_STATS tells the two cache topologies apart:
//!
//!   phase 1 — two client processes -> ONE SuperServer shared cache
//!             (coherency by shared memory; almost no physical I/O)
//!   phase 2 — two EMBEDDED engine processes with PRIVATE caches over one
//!             file (a ServerMode=SuperClassic sandbox reached through
//!             with_embedded(); coherency by LCK_bdb page locks + blocking
//!             ASTs — data travels through the disk, so the same workload
//!             turns into hundreds of reads AND writes)
//!
//! The parent only spawns re-invocations of itself; every attachment lives
//! in a child process (each embedded child IS a full engine).  Rows 1 and 2
//! share a data page, so the two writers fight over one page without ever
//! touching one row.
//!
//! Run (see ../README.md):  cargo run --bin page_cache

use rsfbclient::{prelude::*, FbError, SimpleConnection, SimpleTransaction};
use std::process::{Child, Command};

const SRV_DB: &str = "/tmp/fbhandson/page_cache_rust.fdb"; // server-side path
const EMB_DIR: &str = "/tmp/fbhandson_rust_emb"; // ours, not the server's
const EMB_DB: &str = "/tmp/fbhandson_rust_emb/page_cache_rust.fdb";
const SANDBOX: &str = "/tmp/fbhandson_rust_emb/fbroot-superclassic";
const ROUNDS: usize = 300;

// "srv" attaches through the server; "emb" runs a whole engine in-process
// (the child was spawned with FIREBIRD pointing at the SuperClassic sandbox).
fn attach(mode: &str, create: bool) -> Result<SimpleConnection, FbError> {
    if mode == "srv" {
        let b = rsfbclient::builder_native()
            .with_dyn_link()
            .with_remote()
            .host("localhost")
            .db_name(SRV_DB)
            .user(fb_handson_rust::user())
            .pass(fb_handson_rust::password())
            .clone();
        match b.connect() {
            Ok(c) => Ok(c.into()),
            Err(e) if create => match b.create_database() {
                Ok(c) => Ok(c.into()),
                Err(_) => Err(e),
            },
            Err(e) => Err(e),
        }
    } else {
        let mut b = rsfbclient::builder_native().with_dyn_link().with_embedded();
        b.db_name(EMB_DB).user(fb_handson_rust::user());
        match b.connect() {
            Ok(c) => Ok(c.into()),
            Err(e) if create => match b.create_database() {
                Ok(c) => Ok(c.into()),
                Err(_) => Err(e),
            },
            Err(e) => Err(e),
        }
    }
}

fn tr_config() -> TransactionConfiguration {
    TransactionConfiguration {
        isolation: TrIsolationLevel::ReadCommited(TrRecordVersion::RecordVersion),
        lock_resolution: TrLockResolution::Wait(None),
        ..TransactionConfiguration::default()
    }
}

fn init_db(mode: &str) -> Result<(), FbError> {
    // --init: fresh table, two rows (they will share one data page)
    let mut conn = attach(mode, true)?;
    let mut tr = SimpleTransaction::new(&mut conn, tr_config())?;
    let _ = tr.execute("drop table t", ());
    tr.commit_retaining()?;
    tr.execute("create table t (id int primary key, v int)", ())?;
    tr.commit_retaining()?;
    tr.execute("insert into t values (1, 0)", ())?;
    tr.execute("insert into t values (2, 0)", ())?;
    tr.commit()?;
    Ok(())
}

fn worker(mode: &str, row_id: &str) -> Result<(), FbError> {
    let mut conn = attach(mode, false)?;
    for _ in 0..ROUNDS {
        let mut tr = SimpleTransaction::new(&mut conn, tr_config())?;
        tr.execute(
            &format!("update t set v = v + 1 where id = {}", row_id),
            (),
        )?;
        tr.commit()?;
    }
    let mut tr = SimpleTransaction::new(&mut conn, tr_config())?;
    let (fetches, reads, writes): (i64, i64, i64) = tr
        .query_first(
            "select MON$PAGE_FETCHES, MON$PAGE_READS, MON$PAGE_WRITES \
             from MON$IO_STATS join MON$ATTACHMENTS using (MON$STAT_ID) \
             where MON$ATTACHMENT_ID = CURRENT_CONNECTION",
            (),
        )?
        .unwrap();
    println!(
        "  worker pid {:<6} row {}: {} commits | page fetches={:<6} reads={:<4} writes={}",
        std::process::id(),
        row_id,
        ROUNDS,
        fetches,
        reads,
        writes
    );
    tr.commit()?;
    Ok(())
}

fn check(mode: &str) -> Result<(), FbError> {
    // --check: any lost updates?
    let mut conn = attach(mode, false)?;
    let mut tr = SimpleTransaction::new(&mut conn, tr_config())?;
    let rows: Vec<(i64, i64)> = tr.query("select id, v from t order by id", ())?;
    for (id, v) in rows {
        println!("  final: id={} v={} (expected {})", id, v, ROUNDS);
    }
    tr.commit()?;
    Ok(())
}

fn spawn_child(args: &[&str], firebird_root: Option<&str>) -> Result<Child, FbError> {
    let mut cmd = Command::new(std::env::current_exe()?);
    cmd.args(args);
    if let Some(root) = firebird_root {
        cmd.env("FIREBIRD", root);
    }
    Ok(cmd.spawn()?)
}

fn run_phase(mode: &str, firebird_root: Option<&str>) -> Result<(), FbError> {
    spawn_child(&["--init", mode], firebird_root)?.wait()?;
    let mut w1 = spawn_child(&["--worker", mode, "1"], firebird_root)?;
    let mut w2 = spawn_child(&["--worker", mode, "2"], firebird_root)?;
    w1.wait()?;
    w2.wait()?;
    spawn_child(&["--check", mode], firebird_root)?.wait()?;
    Ok(())
}

// Build the embedded sandbox: a FIREBIRD root whose firebird.conf says
// SuperClassic, so each process locks the file SHARED and runs its own
// page cache (see 'Three layers of arbitration').
fn make_sandbox() -> Result<(), FbError> {
    std::fs::create_dir_all(SANDBOX)?;
    for f in ["plugins", "intl", "tzdata", "firebird.msg", "security6.fdb"] {
        let _ = std::os::unix::fs::symlink(
            format!("/opt/firebird/{}", f),
            format!("{}/{}", SANDBOX, f),
        );
    }
    std::fs::write(
        format!("{}/firebird.conf", SANDBOX),
        "ServerMode = SuperClassic\n",
    )?;
    let _ = std::fs::remove_file(EMB_DB);
    Ok(())
}

fn main() -> Result<(), FbError> {
    let args: Vec<String> = std::env::args().collect();
    match args.get(1).map(String::as_str) {
        Some("--init") => return init_db(&args[2]),
        Some("--worker") => return worker(&args[2], &args[3]),
        Some("--check") => return check(&args[2]),
        _ => {}
    }

    println!("phase 1: two client processes, ONE SuperServer shared cache");
    run_phase("srv", None)?;

    std::fs::create_dir_all(EMB_DIR)?;
    make_sandbox()?;

    println!("phase 2: two EMBEDDED engine processes, PRIVATE page caches");
    run_phase("emb", Some(SANDBOX))?;
    println!("same workload — the private caches paid for coherency in disk I/O.");
    Ok(())
}
